<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/assignment/ass6/hpdp/303/Assignment6_Group303.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Assignment 6: Mastering Big Data Handling**

#### Group Members:
|Name|Matric No.|
|----------|------|
|ANG YI QIN |A21EC0163|
|SOO WAN YING |A21EC0227|



### **Data Selection**
This dataset provides comprehensive information on airline flights, specifically focusing on delays and cancellations during the year 2019. This dataset is accessible on Kaggle through the link [here](https://www.kaggle.com/datasets/threnjen/2019-airline-delays-and-cancellations).

The dataset contains a substantial number of entries, with a RangeIndex spanning from 0 to 7,213,445. A total of 27 columns provide detailed information about each flight record. The dataset includes columns such as FL_DATE, OP_CARRIER, CRS_DEP_TIME, DEP_DELAY, CANCELLED, and many more. Data types include integers, floats, and categories, indicating a mix of numerical and categorical information. Essential attributes include flight date, carrier information, departure and arrival times, delays, cancellation status, and various delay types (carrier delay, weather delay, etc.).

### **Loading The Datasets**

In [1]:
# Install opendatasets to read the dataset from kaggle
!pip install opendatasets

In [2]:
# Import required libraries
import pandas as pd
import opendatasets as od
import warnings
warnings.filterwarnings('ignore')

# Import dataset
od.download("https://www.kaggle.com/datasets/threnjen/2019-airline-delays-and-cancellations/")

file = ('full_data_flightdelay.csv')

Skipping, found downloaded files in "./2019-airline-delays-and-cancellations" (use force=True to force download)


In [3]:
from google.colab import files

# Load the dataset
file_path = '/content/2019-airline-delays-and-cancellations/full_data_flightdelay.csv'
df = pd.read_csv(file_path, encoding='ISO-8859-1')

In [4]:
# Display the first few rows of the dataframe
df.head()

MONTH  DAY_OF_WEEK  DEP_DEL15 DEP_TIME_BLK  DISTANCE_GROUP  SEGMENT_NUMBER  \
0      1            7          0    0800-0859               2               1   
1      1            7          0    0700-0759               7               1   
2      1            7          0    0600-0659               7               1   
3      1            7          0    0600-0659               9               1   
4      1            7          0    0001-0559               7               1   

   CONCURRENT_FLIGHTS  NUMBER_OF_SEATS            CARRIER_NAME  \
0                  25              143  Southwest Airlines Co.   
1                  29              191    Delta Air Lines Inc.   
2                  27              199    Delta Air Lines Inc.   
3                  27              180    Delta Air Lines Inc.   
4                  10              182        Spirit Air Lines   

   AIRPORT_FLIGHTS_MONTH  ...  PLANE_AGE       DEPARTING_AIRPORT  LATITUDE  \
0                  13056  ...          8  McCarran International     36.08   
1                  13056  ...          3  McCarran International     36.08   
2                  13056  ...         18  McCarran International     36.08   
3                  13056  ...          2  McCarran International     36.08   
4                  13056  ...          1  McCarran International     36.08   

   LONGITUDE  PREVIOUS_AIRPORT  PRCP  SNOW SNWD  TMAX  AWND  
0   -115.152              NONE   0.0   0.0  0.0  65.0  2.91  
1   -115.152              NONE   0.0   0.0  0.0  65.0  2.91  
2   -115.152              NONE   0.0   0.0  0.0  65.0  2.91  
3   -115.152              NONE   0.0   0.0  0.0  65.0  2.91  
4   -115.152              NONE   0.0   0.0  0.0  65.0  2.91  

[5 rows x 26 columns]

In [5]:
# # Display descriptive statistics of the dataframe
df.describe()

MONTH   DAY_OF_WEEK     DEP_DEL15  DISTANCE_GROUP  \
count  6.489062e+06  6.489062e+06  6.489062e+06    6.489062e+06   
mean   6.607062e+00  3.935598e+00  1.891441e-01    3.821102e+00   
std    3.396853e+00  1.995200e+00  3.916231e-01    2.382233e+00   
min    1.000000e+00  1.000000e+00  0.000000e+00    1.000000e+00   
25%    4.000000e+00  2.000000e+00  0.000000e+00    2.000000e+00   
50%    7.000000e+00  4.000000e+00  0.000000e+00    3.000000e+00   
75%    1.000000e+01  6.000000e+00  0.000000e+00    5.000000e+00   
max    1.200000e+01  7.000000e+00  1.000000e+00    1.100000e+01   

       SEGMENT_NUMBER  CONCURRENT_FLIGHTS  NUMBER_OF_SEATS  \
count    6.489062e+06        6.489062e+06     6.489062e+06   
mean     3.046890e+00        2.783675e+01     1.337397e+02   
std      1.757864e+00        2.151060e+01     4.645213e+01   
min      1.000000e+00        1.000000e+00     4.400000e+01   
25%      2.000000e+00        1.100000e+01     9.000000e+01   
50%      3.000000e+00        2.300000e+01     1.430000e+02   
75%      4.000000e+00        3.900000e+01     1.720000e+02   
max      1.500000e+01        1.090000e+02     3.370000e+02   

       AIRPORT_FLIGHTS_MONTH  AIRLINE_FLIGHTS_MONTH  \
count           6.489062e+06           6.489062e+06   
mean            1.268458e+04           6.296058e+04   
std             8.839796e+03           3.438223e+04   
min             1.100000e+03           5.582000e+03   
25%             5.345000e+03           2.503400e+04   
50%             1.156200e+04           7.087800e+04   
75%             1.761500e+04           8.631200e+04   
max             3.525600e+04           1.177280e+05   

       AIRLINE_AIRPORT_FLIGHTS_MONTH  ...  FLT_ATTENDANTS_PER_PASS  \
count                   6.489062e+06  ...             6.489062e+06   
mean                    3.459251e+03  ...             9.753707e-05   
std                     4.251139e+03  ...             8.644459e-05   
min                     1.000000e+00  ...             0.000000e+00   
25%                     6.540000e+02  ...             3.419267e-05   
50%                     2.251000e+03  ...             6.178236e-05   
75%                     4.806000e+03  ...             1.441659e-04   
max                     2.183700e+04  ...             3.484077e-04   

       GROUND_SERV_PER_PASS     PLANE_AGE      LATITUDE     LONGITUDE  \
count          6.489062e+06  6.489062e+06  6.489062e+06  6.489062e+06   
mean           1.355612e-04  1.153211e+01  3.670581e+01 -9.425515e+01   
std            4.649970e-05  6.935706e+00  5.500804e+00  1.790952e+01   
min            7.134695e-06  0.000000e+00  1.844000e+01 -1.593460e+02   
25%            9.889412e-05  5.000000e+00  3.343600e+01 -1.063770e+02   
50%            1.246511e-04  1.200000e+01  3.750500e+01 -8.790600e+01   
75%            1.772872e-04  1.700000e+01  4.077900e+01 -8.093600e+01   
max            2.289855e-04  3.200000e+01  6.116900e+01 -6.600200e+01   

               PRCP          SNOW          SNWD          TMAX          AWND  
count  6.489062e+06  6.489062e+06  6.489062e+06  6.489062e+06  6.489062e+06  
mean   1.037063e-01  3.159310e-02  9.152397e-02  7.146846e+01  8.341329e+00  
std    3.432134e-01  3.170163e-01  7.281285e-01  1.835333e+01  3.607604e+00  
min    0.000000e+00  0.000000e+00  0.000000e+00 -1.000000e+01  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  5.900000e+01  5.820000e+00  
50%    0.000000e+00  0.000000e+00  0.000000e+00  7.400000e+01  7.830000e+00  
75%    2.000000e-02  0.000000e+00  0.000000e+00  8.600000e+01  1.029000e+01  
max    1.163000e+01  1.720000e+01  2.520000e+01  1.150000e+02  3.378000e+01  

[8 rows x 22 columns]

In [6]:
# Display information about the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6489062 entries, 0 to 6489061
Data columns (total 26 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   MONTH                          int64  
 1   DAY_OF_WEEK                    int64  
 2   DEP_DEL15                      int64  
 3   DEP_TIME_BLK                   object 
 4   DISTANCE_GROUP                 int64  
 5   SEGMENT_NUMBER                 int64  
 6   CONCURRENT_FLIGHTS             int64  
 7   NUMBER_OF_SEATS                int64  
 8   CARRIER_NAME                   object 
 9   AIRPORT_FLIGHTS_MONTH          int64  
 10  AIRLINE_FLIGHTS_MONTH          int64  
 11  AIRLINE_AIRPORT_FLIGHTS_MONTH  int64  
 12  AVG_MONTHLY_PASS_AIRPORT       int64  
 13  AVG_MONTHLY_PASS_AIRLINE       int64  
 14  FLT_ATTENDANTS_PER_PASS        float64
 15  GROUND_SERV_PER_PASS           float64
 16  PLANE_AGE                      int64  
 17  DEPARTING_AIRPORT              object 
 18  LA

In [7]:
# Check for duplicates
print(df.duplicated().sum())

28473


In [8]:
duplicate_rows = df[df.duplicated()]
print(duplicate_rows.head())

    MONTH  DAY_OF_WEEK  DEP_DEL15 DEP_TIME_BLK  DISTANCE_GROUP  \
44      1            7          0    0700-0759               5   
46      1            7          0    0700-0759               5   
51      1            7          0    0800-0859               1   
73      1            7          0    0700-0759               6   
85      1            7          0    0700-0759               6   

    SEGMENT_NUMBER  CONCURRENT_FLIGHTS  NUMBER_OF_SEATS  \
44               1                  29              129   
46               1                  29              129   
51               1                  25              143   
73               1                  29              129   
85               1                  29              129   

              CARRIER_NAME  AIRPORT_FLIGHTS_MONTH  ...  PLANE_AGE  \
44           Allegiant Air                  13056  ...         11   
46           Allegiant Air                  13056  ...         11   
51  Southwest Airlines Co.               

In [9]:
# Check missing data
df.isna().sum()

MONTH                            0
DAY_OF_WEEK                      0
DEP_DEL15                        0
DEP_TIME_BLK                     0
DISTANCE_GROUP                   0
SEGMENT_NUMBER                   0
CONCURRENT_FLIGHTS               0
NUMBER_OF_SEATS                  0
CARRIER_NAME                     0
AIRPORT_FLIGHTS_MONTH            0
AIRLINE_FLIGHTS_MONTH            0
AIRLINE_AIRPORT_FLIGHTS_MONTH    0
AVG_MONTHLY_PASS_AIRPORT         0
AVG_MONTHLY_PASS_AIRLINE         0
FLT_ATTENDANTS_PER_PASS          0
GROUND_SERV_PER_PASS             0
PLANE_AGE                        0
DEPARTING_AIRPORT                0
LATITUDE                         0
LONGITUDE                        0
PREVIOUS_AIRPORT                 0
PRCP                             0
SNOW                             0
SNWD                             0
TMAX                             0
AWND                             0
dtype: int64

### **Strategies for Big Datsets**
Apply five smart strategies to handle large datasets effectively

##### Load Less Data: Strategically load only the essential portions of the dataset to optimize memory usage.

Chooses specific columns (like "CARRIER_NAME," "DEPARTING_AIRPORT," etc.) from a dataset that contains information about airline delays and cancellations. Loads only the selected columns from a large dataset (CSV file) into a smaller DataFrame using the Pandas library.

In [11]:
# Specify the columns want to load
selected_columns = ["CARRIER_NAME", "DEPARTING_AIRPORT", "PLANE_AGE", "CONCURRENT_FLIGHTS", "LATITUDE", "LONGITUDE"]

# Load only the essential portions of the dataset
file_path = '/content/2019-airline-delays-and-cancellations/full_data_flightdelay.csv'
less_data_df = pd.read_csv(file_path, usecols = selected_columns)

# Explore the loaded data
print(less_data_df.head())

   CONCURRENT_FLIGHTS            CARRIER_NAME  PLANE_AGE  \
0                  25  Southwest Airlines Co.          8   
1                  29    Delta Air Lines Inc.          3   
2                  27    Delta Air Lines Inc.         18   
3                  27    Delta Air Lines Inc.          2   
4                  10        Spirit Air Lines          1   

        DEPARTING_AIRPORT  LATITUDE  LONGITUDE  
0  McCarran International     36.08   -115.152  
1  McCarran International     36.08   -115.152  
2  McCarran International     36.08   -115.152  
3  McCarran International     36.08   -115.152  
4  McCarran International     36.08   -115.152  


In [25]:
print("Memory usage for load less data:", less_data_df.memory_usage().sum() / (1024 ** 2), "MB")

import timeit
time = timeit.timeit(lambda:less_data_df, number=1)
print("Computation time for load less data:", time, "seconds")

Memory usage for load less data: 297.0458068847656 MB
Computation time for load less data: 1.7610000213608146e-06 seconds


#### Use Chunking: Process the data in smaller pieces to avoid memory issues.


This code reads a large CSV file in chunks of 10000 rows each, and saves each chunk to a separate CSV file named 'chunk1.csv', 'chunk2.csv', and so on. It's a memory-efficient way to handle large datasets. The garbage collector is used to manage memory during the process.

In [23]:
import gc

chunk_size = 10000
num = 1
chunk_result_df = pd.DataFrame()
total_memory_usage = 0

file_path = '/content/2019-airline-delays-and-cancellations/full_data_flightdelay.csv'

chunks = pd.read_csv(file_path, chunksize=chunk_size)

# Start measuring time
chunk_start_time = timeit.default_timer()

for chunk in chunks:
    if num <= 5:
      print(f"Chunk {num}:\n", chunk.head())
    chunk_memory_usage = chunk.memory_usage(deep=True).sum() / (1024 ** 2)
    total_memory_usage += chunk_memory_usage
    chunk.to_csv('chunk'+str(num)+'.csv', index=False)
    gc.collect()
    num+=1

# Stop measuring time
chunk_end_time = timeit.default_timer()
chunk_elapsed_time = chunk_end_time - chunk_start_time

Chunk 1:
    MONTH  DAY_OF_WEEK  DEP_DEL15 DEP_TIME_BLK  DISTANCE_GROUP  SEGMENT_NUMBER  \
0      1            7          0    0800-0859               2               1   
1      1            7          0    0700-0759               7               1   
2      1            7          0    0600-0659               7               1   
3      1            7          0    0600-0659               9               1   
4      1            7          0    0001-0559               7               1   

   CONCURRENT_FLIGHTS  NUMBER_OF_SEATS            CARRIER_NAME  \
0                  25              143  Southwest Airlines Co.   
1                  29              191    Delta Air Lines Inc.   
2                  27              199    Delta Air Lines Inc.   
3                  27              180    Delta Air Lines Inc.   
4                  10              182        Spirit Air Lines   

   AIRPORT_FLIGHTS_MONTH  ...  PLANE_AGE       DEPARTING_AIRPORT  LATITUDE  \
0                  13056  ..

In [24]:
print(f"\nTotal memory usage for all chunks: {total_memory_usage:.2f} MB")
print(f"Total elapsed time for processing chunks: {chunk_elapsed_time:.2f} seconds")


Total memory usage for all chunks: 2955.80 MB
Total elapsed time for processing chunks: 208.50 seconds


#### Optimize Data Types: Fine-tune data types to maximize efficiency and minimize memory consumption.


Converting specific data types to more memory-efficient alternatives, and displays data types and memory usage before and after optimization. The optimization involves converting float64 to float16, int64 to int16, object to category.

In [15]:
# Display data types and memory usage before optimization
print("Data types before optimization:")
print(df.dtypes)
print("\nMemory usage before optimization:")
print(df.memory_usage(deep=True).sum() / (1024 * 1024), "MB")

# Optimize data types
for col in df.columns:
    if df[col].dtype == 'float64':
        df[col] = df[col].astype('float16')
    if df[col].dtype == 'int64':
        df[col] = df[col].astype('int16')
    if df[col].dtype == 'object':
        df[col] = df[col].astype('category')

# Display data types and memory usage after optimization
print("\nData types after optimization:")
print(df.dtypes)
print("\nMemory usage after optimization:")
print(df.memory_usage(deep=True).sum() / (1024 * 1024), "MB")

Data types before optimization:
MONTH                              int64
DAY_OF_WEEK                        int64
DEP_DEL15                          int64
DEP_TIME_BLK                      object
DISTANCE_GROUP                     int64
SEGMENT_NUMBER                     int64
CONCURRENT_FLIGHTS                 int64
NUMBER_OF_SEATS                    int64
CARRIER_NAME                      object
AIRPORT_FLIGHTS_MONTH              int64
AIRLINE_FLIGHTS_MONTH              int64
AIRLINE_AIRPORT_FLIGHTS_MONTH      int64
AVG_MONTHLY_PASS_AIRPORT           int64
AVG_MONTHLY_PASS_AIRLINE           int64
FLT_ATTENDANTS_PER_PASS          float64
GROUND_SERV_PER_PASS             float64
PLANE_AGE                          int64
DEPARTING_AIRPORT                 object
LATITUDE                         float64
LONGITUDE                        float64
PREVIOUS_AIRPORT                  object
PRCP                             float64
SNOW                             float64
SNWD                     

In [27]:
print("\nMemory usage for optimized data types:", df.memory_usage(deep=True).sum() / (1024 * 1024), "MB")

# Measure computation time on the optimized DataFrame
import timeit
time = timeit.timeit(lambda: df, number=1)
print("Computation time for optimize data types:", time, "seconds")


Memory usage for optimized data types: 303.28170013427734 MB
Computation time for optimize data types: 1.4990000636316836e-06 seconds


#### Sampling: Implement sampling methodologies to extract meaningful insights from a subset of the dataset.


Randomly selects 10% of rows from the original DataFrame (df) using the sample method. Then, prints the sizes (number of rows and columns) of both the original and sampled datasets, providing an overview of the subset compared to the complete dataset. The use of random_state=42 ensures reproducibility.

In [17]:
# Sample a fraction of the data (e.g., 10%)
sample_df = df.sample(frac=0.1, random_state=42)

# Display information about the sampled data
print("Original dataset size:", df.shape)
print("Sampled dataset size:", sample_df.shape)

# Display first few rows of the sampled data
print(sample_df.head())

Original dataset size: (6489062, 26)
Sampled dataset size: (648906, 26)
         MONTH  DAY_OF_WEEK  DEP_DEL15 DEP_TIME_BLK  DISTANCE_GROUP  \
984735       3            5          0    1500-1559               3   
4740332      9            7          0    2200-2259               5   
6361351     12            1          0    1900-1959               3   
4012189      8            3          0    1700-1759               6   
4789024      9            5          0    0800-0859               2   

         SEGMENT_NUMBER  CONCURRENT_FLIGHTS  NUMBER_OF_SEATS  \
984735                4                  27               69   
4740332               1                   7              128   
6361351               5                  15               76   
4012189               2                  22              276   
4789024               1                   3               76   

                   CARRIER_NAME  AIRPORT_FLIGHTS_MONTH  ...  PLANE_AGE  \
984735    Midwest Airline, Inc.           

In [28]:
print("Memory usage for sampling:", sample_df.memory_usage(deep=True).sum() / (1024 ** 2), "MB")

# Measure computation time on the sampled DataFrame
import timeit
time = timeit.timeit(lambda: sample_df, number=1)
print("Computation time for sampling:", time, "seconds")

Memory usage for sampling: 35.32160568237305 MB
Computation time for sampling: 1.7969996406463906e-06 seconds


#### Parallelize with Dask: Dask is a powerful library that extends pandas to enable parallel and distributed computing. It's particularly useful for handling larger-than-memory datasets.

Converts a Pandas DataFrame (df) into a Dask DataFrame (dask_df) with 10 partitions. Computes the mean of the 'DISTANCE_GROUP' column using Dask. Executes the computation and prints the computed mean distance. Performs a parallelized groupby operation on 'DEP_TIME_BLK' and calculates the sum of 'NUMBER_OF_SEATS' for each group.

In [19]:
!pip install dask

In [20]:
import dask
import dask.dataframe as dd

# Assuming 'df' is your Pandas DataFrame
dask_df = dd.from_pandas(df, npartitions=10)  # Adjust the number of partitions as needed

# Calculate the mean of a column
mean_distance = dask_df['DISTANCE_GROUP'].mean()

# Compute the result
result_mean_distance = mean_distance.compute()
print("Mean Distance:", result_mean_distance)

# Perform a parallelized groupby operation
result_groupby = dask_df.groupby('DEP_TIME_BLK')['NUMBER_OF_SEATS'].sum().compute()
print("\nGrouped by DEP_TIME_BLK, Sum of NUMBER_OF_SEATS:")
print(result_groupby)

Mean Distance: 3.8211020329286423

Grouped by DEP_TIME_BLK, Sum of NUMBER_OF_SEATS:
DEP_TIME_BLK
0001-0559    21296152
0600-0659    60729846
0700-0759    60806350
0800-0859    60626095
0900-0959    52094550
1000-1059    52217265
1100-1159    52471423
1200-1259    51546305
1300-1359    46913175
1400-1459    48957244
1500-1559    50397242
1600-1659    48821278
1700-1759    53445439
1800-1859    49489393
1900-1959    48311233
2000-2059    44202301
2100-2159    30808800
2200-2259    26156996
2300-2359     8554372
Name: NUMBER_OF_SEATS, dtype: int64


In [29]:
# Measure memory usage for the Dask DataFrame
print("Memory usage for parallelize with Dask:", dask_df.memory_usage(deep=True).sum().compute() / (1024 ** 2), "MB")

# Measure computation time on the Dask DataFrame
import timeit
time = timeit.timeit(lambda: dask_df.compute(), number=1)
print("Computation time for parallelize with Dask:", time, "seconds")

Memory usage for parallelize with Dask: 303.70980644226074 MB
Computation time for parallelize with Dask: 1.8671546749992558 seconds


### Comparative Analysis

**Result and analysis:**

| **Strategy**                | **Memory Usage**      | **Computation Time**  | **Analysis**                                                                                        |
|-----------------------------|-----------------------|-----------------------|-----------------------------------------------------------------------------------------------------|
| **Loading Less Data**       | 297.05 MB             | 1.76 microseconds     | Efficient with low memory usage, fast loading. |
| **Processing Chunks**       | 2955.80 MB (Total)    | 208.50 seconds (Total)| High memory usage, slow processing. Processing the entire dataset in chunks may not be efficient.   |
| **Optimizing Data Types**   | 303.28 MB             | 1.50 microseconds     | Slight increase in memory usage after optimization, but significant reduction in computation time.   |
| **Sampling**                | 35.32 MB              | 1.80 microseconds     | Very low memory usage and fast computation time. |
| **Parallelizing with Dask** | 303.71 MB             | 1.87 seconds          | Similar memory usage to optimized data types, but faster computation time. |




**Comparision between five strategies:**

| **Strategy**               | **Traditional Approach**                         | **Advanced Approach**                          | **Key Benefits**                                                               |
|-----------------------------|--------------------------------------------------|----------------------------------------------|---------------------------------------------------------------------------------|
| **1. Load Less Data**       | Loads entire data into memory                    | Loads only essential parts strategically      | Greatly improves handling of large data, making it possible to analyze big datasets |
| **2. Use Chunking**         | Processes all data in one go                     | Breaks data into smaller, manageable parts     | Makes data processing smoother, reducing memory issues and allowing for parallel processing |
| **3. Optimize Data Types**  | Automatically infers data types                   | Fine-tunes data types for efficiency           | Reduces memory usage significantly by using more compact data types            |
| **4. Sampling**             | Analyzes the entire dataset                       | Analyzes a representative subset (e.g., 10%) | Speeds up analysis, especially when a full-scale examination is not needed        |
| **5. Parallelize with Dask**| Lacks support for parallel computing             | Uses Dask for parallel and distributed computing | Drastically cuts down computation time, enables working with datasets larger than memory |




### Conclusion

1. If memory usage is a critical factor, consider optimizing data types or sampling.
2. Processing chunks may not be efficient in terms of memory usage or processing time.
3. Optimizing data types reduces memory footprint with minimal impact on computation time.
4. Sampling is effective for quick exploration on a smaller dataset.
5. Parallelizing with Dask is suitable for large-scale parallel processing, balancing memory usage and computation time.

In summary, the implementation of advanced strategies in handling big data, including loading less data, utilizing chunking, optimizing data types, sampling, and Parallelize with Dask, offers substantial advantages. These approaches lead to decreased memory usage, quicker computation times, and enhanced overall efficiency when compared to conventional methods. The capacity to manage larger datasets and optimize resource utilization underscores the significance of these advanced strategies for data scientists seeking meaningful insights from extensive and intricate datasets. Embracing these methods is pivotal for efficient and scalable big data analysis.